pkg/awsutils/awsutils.go

```diff
@@ -674,13 +684,31 @@ func (cache *EC2InstanceMetadataCache) AllocIPAddress(eniID string) error {
 	return nil
 }
 
+// GetENIipLimit return IP address limit per ENI based on EC2 instance type
+func (cache *EC2InstanceMetadataCache) GetENIipLimit() (int64, error) {
+	ipLimit, ok := InstanceIPsAvailable[cache.instanceType]
+	if !ok {
+		log.Errorf("Failed to get eni IP limit due to unknown instance type %s", cache.instanceType)
+		return 0, errors.Errorf("vpc ip resource(eni ip limit): unknown instance thype %s", cache.instanceType)
+	}
+
+	return ipLimit - 1, nil
+}
+
 // AllocAllIPAddress allocates all IP addresses available on eni
 func (cache *EC2InstanceMetadataCache) AllocAllIPAddress(eniID string) error {
 	log.Infof("Trying to allocate all available ip addresses on eni: %s", eniID)
 
+	ipLimit, err := cache.GetENIipLimit()
+
+	if err != nil {
+		// for unknown instance type, will allocate one ip address at a time
+		ipLimit = 1
+	}
+
 	input := &ec2.AssignPrivateIpAddressesInput{
 		NetworkInterfaceId:             aws.String(eniID),
-		SecondaryPrivateIpAddressCount: aws.Int64(1),
+		SecondaryPrivateIpAddressCount: aws.Int64(ipLimit),
 	}
 
 	for {
```

Comment:
Do we still need this in a for loop since we are allocating the maximum in a single request?

Reply:
For the "unknown" instance type, we still need loop, since we are still calling one IP address at at time

Reply:
feels like quite lot of duplicate code when break this into two paths
```
        if err != nil {
                // for unknown instance type, will allocate one ip address at a time
                ipLimit = 1 

                input := &ec2.AssignPrivateIpAddressesInput{
                        NetworkInterfaceId:             aws.String(eniID),
                        SecondaryPrivateIpAddressCount: aws.Int64(ipLimit),
                }   

                for {
                        // until error
                        _, err := cache.ec2SVC.AssignPrivateIpAddresses(input)
                        if err != nil {
                                if containsPrivateIPAddressLimitExceededError(err) {
                                        return nil 
                                }   
                                log.Errorf("Failed to allocate a private IP address %v", err)
                                return errors.Wrap(err, "allocate ip address: failed to allocate a private IP address")
                        }   
                }   
        } else {
                // for known instance type, will allocate max number ip address for that interface
                input := &ec2.AssignPrivateIpAddressesInput{
                        NetworkInterfaceId:             aws.String(eniID),
                        SecondaryPrivateIpAddressCount: aws.Int64(ipLimit),
                }   

                _, err := cache.ec2SVC.AssignPrivateIpAddresses(input)
                if err != nil {
                        if containsPrivateIPAddressLimitExceededError(err) {
                                return nil 
                        }   
                        log.Errorf("Failed to allocate a private IP address %v", err)
                        return errors.Wrap(err, "allocate ip address: failed to allocate a private IP address")
                }   
        } 
```